### PR TITLE
feat: introduce new writer trait 

### PR DESCRIPTION
**Pull request by ZENOTME**
This PR is the 1 step for #243.

It introduces new writer trait and implements it in
1. rolling writer

### CODE REVIEW

**Comment by liurenjie1024** [comment]
Is this ready for review or it's still a wip?

**Comment by ZENOTME** [comment]
> Is this ready for review or it's still a wip? 

Still a wip

**Comment by ZENOTME** [comment]
It's ready to review now. But I just implemented it in rolling writer now to make this PR smaller.

File: icelake/src/io/mod.rs [hunk 1]
```diff
@@ -14,3 +13,26 @@ pub mod parquet;
...
+    async fn close(&mut self) -> Result<Vec<DataFileBuilder>>;
+}
+
+#[async_trait::async_trait]
+pub trait RecordBatchWriterBuilder: Send + Sync + Clone + 'static {
+    type R: RecordBatchWriter;
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R>;
+}
+
+pub trait SingletonWriterStatus {
```
Comment (liurenjie1024) [inline icelake/src/io/mod.rs:35]: This usage is a little weird to me, how about this:
```rust
pub trait SingleFileWriter: RecordBatchWriter {
   fn current_file(&self) -> String;
  fn current_row_num(&self) -> usize;
}
```

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM

**Review by liurenjie1024 — APPROVED** [review]

File: icelake/src/io/mod.rs (hunk 1 above)
Comment (ZENOTME) [reply icelake/src/io/mod.rs:35]: nice!

File: icelake/src/io/writer_builder.rs [hunk 2]
```diff
@@ -83,109 +61,82 @@ impl<L: FileAppenderLayer<DefaultFileAppender>> WriterBuilder<L> {
...
-        .await
+    pub async fn build_append_only_writer<B: RecordBatchWriterBuilder>(
+        self,
+        builder: B,
+    ) -> Result<AppendOnlyWriter<B>> {
+        AppendOnlyWriter::try_new(self.table_metadata, builder).await
     }
 
-    pub async fn build_upsert_writer(
+    pub async fn build_upsert_writer<B: RecordBatchWriterBuilder>(
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply icelake/src/io/writer_builder.rs:127]: For now, Upsert is 
```
enum {
  Partition(Partition< EqualityDeltaWriter >),
  Unpartition(EqualityDeltaWriter) 
}
```
I will separate it out as more general writer such as following later, so we can remove the concept of UpsertWriter.
```
enum DispatcherWriter<L,R>;
struct PartitionWriter<InnerWriter>;
struct EqualityDeltaWriter;
```

File: icelake/src/types/partition_splitter.rs [hunk 3]
```diff
@@ -39,15 +39,18 @@ impl FieldProjector {
...
-                .metadata()
-                .get(COLUMN_ID_META_KEY)
-                .expect("column_id must be set")
-                .parse()
-                .expect("column_id must can be parse as i64");
+            let id: i32 = if let Some(id) = field.metadata().get(COLUMN_ID_META_KEY) {
+                if let Ok(id) = id.parse::<i32>() {
+                    id
+                } else {
+                    continue;
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply icelake/src/types/partition_splitter.rs]: Modify it in next PR.

**Comment by liurenjie1024** [comment]
LGTM, let's move on.

File: icelake/src/io/appender/rolling_writer.rs [hunk 4]
```diff
@@ -33,37 +72,38 @@ pub struct RollingWriter {
 
     result: Vec<DataFileBuilder>,
 
-    table_config: TableConfigRef,
+    config: RollingWriterConfig,
+    parquet_config: ParquetWriterConfig,
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply icelake/src/io/appender/rolling_writer.rs:76]: We can separate the file writer out. Using
```
struct RollingWriter<F>;
struct ParquetWriter;
struct OrcWriter;
```

**Pull request merged**